### ISSUE TITLE
Fix wind downscaling under Iris 2.1

### ISSUE THREAD

**Issue by benfitzpatrick** [metoppv/improver#669]
On our latest Python 3 stack (Iris 2.1, Python 3.6.6, Dask 0.18.1) we have an exception when running `improver wind-downscaling`:

```
Traceback (most recent call last):
  File "/opt/improver/bin/improver-wind-downscaling", line 156, in <module>
    main()
  File "/opt/improver/bin/improver-wind-downscaling", line 133, in main
    wind_speed = wind_speed_list.merge_cube()
  File "/lib/python3.6/site-packages/iris/cube.py", line 382, in merge_cube
    merged_cube, = proto_cube.merge()
  File "/lib/python3.6/site-packages/iris/_merge.py", line 1241, in merge
    merged_data = as_concrete_data(merged_data)
  File "/lib/python3.6/site-packages/iris/_lazy_data.py", line 156, in as_concrete_data
    data, = _co_realise_lazy_arrays([data])
  File "/lib/python3.6/site-packages/iris/_lazy_data.py", line 120, in _co_realise_lazy_arrays
    computed_arrays = da.compute(*arrays)
  File "/lib/python3.6/site-packages/dask/base.py", line 402, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/lib/python3.6/site-packages/dask/threaded.py", line 75, in get
    pack_exception=pack_exception, **kwargs)
  File "/lib/python3.6/site-packages/dask/local.py", line 521, in get_async
    raise_exception(exc, tb)
  File "/lib/python3.6/site-packages/dask/compatibility.py", line 69, in reraise
    raise exc
  File "/lib/python3.6/site-packages/dask/local.py", line 290, in execute_task
    result = _execute_task(task, data)
  File "/lib/python3.6/site-packages/dask/local.py", line 270, in _execute_task
    args2 = [_execute_task(a, cache) for a in args]
  File "/lib/python3.6/site-packages/dask/local.py", line 270, in <listcomp>
    args2 = [_execute_task(a, cache) for a in args]
  File "/lib/python3.6/site-packages/dask/local.py", line 271, in _execute_task
    return func(*args2)
TypeError: op_getitem expected 2 arguments, got 4
```

If we take a copy of dask, point to that, and hackily alter it to print out what its doing:
```diff
--- /lib/python3.6/site-packages/dask/local.py	2018-06-23 15:36:48.000000000 +0000
+++ /opt/improver/lib/dask/local.py	2018-09-12 12:47:09.000000000 +0000
@@ -512,12 +512,18 @@
                 key, res_info, failed = queue_get(queue)
                 if failed:
                     exc, tb = loads(res_info)
-                    if rerun_exceptions_locally:
+                    if rerun_exceptions_locally or True:
                         data = dict((dep, state[cache][dep])
                                     for dep in get_dependencies(dsk, key))
                         task = dsk[key]
+                        print("Tried to run", task, data, key, res_info)
                         _execute_task(task, data)  # Re-execute locally
```

then we get something like:

```
Tried to run (<built-in function getitem>, (<built-in function getitem>, array-original-somehexkey, (slice(0, 16, None), slice(0, 970, None), slice(0, 1042, None)), False, False), (None, slice(None, None, None), slice(None, None, None), slice(None, None, None))) {array-original-somehexkey: array([[[9.00655937e+00, 8.97640610e+00, 8.96419144e+00, ...,
[edited out lines]
 1.05287504e+01, 1.05018816e+01, 1.04815016e+01]]], dtype=float32)} (array-stack-someotherhexkey, 0, 0, 0, 0) (TypeError(op_getitem expected 2 arguments, got 4,), <traceback object at 0xaaaaaaa>)
```

print statements in just above the lines mentioned in the traceback in is_concrete_data and _co_realise_lazy_arrays:

something like:
```
print("is_concrete_data", data, data.has_lazy_data())
```

and
```
print("_co_realise_lazy_arrays", arrays)
```
gave:

```
is_concrete_data dask.array<stack, shape=(12, 33, 970, 1042), dtype=float32, chunksize=(1, 16, 970, 1042)> True
_co_realise_lazy_arrays [dask.array<stack, shape=(12, 33, 970, 1042), dtype=float32, chunksize=(1, 16, 970, 1042)>]
```

Acceptance criteria:
See GitHub ticket.

**Comment by LaurenceBeard** [comment]
This problem is hard to reproduce, but can be done by pickling the data pre-merge. 

It seems that within merge, the problem is caused by stacking the arrays using dask (making them lazy, then realising them again), it seems to be particular to the dimensionality of the problem, and can only be replicated as a minimal example with similarly large arrays (eg. particularly slices with more than 16 levels), so the problem seems to lie fundamentally with the chunking of the numpy arrays.

Potential solutions:

•	taking smaller sub-slices in either vertical and/or horizontal coordinates (merging twice, rather than once)
•	an alternative to iris.merge that doesn't use dask (namely replacing da.stack with np.stack in iris' multidim_lazy_stack)
•	a manual merge in this case, not using iris' merge, just naively stacking the results and popping out the relevant dim (realization in this case), effectively a concatenate.

None of these are particular attractive solutions, but serve as a sticking plaster until there is an upstream fix (it appears to be fundamentally a dask bug, so could take some time to fix and trickle down into our stack)

**Comment by LaurenceBeard** [comment]
Iris chooses an arbitrary number of points (inherited from biggus), over which it reduces the chunksize.
```
# A magic value, chosen to minimise chunk creation time and chunk processing
# time within dask.
_MAX_CHUNK_SIZE = 8 * 1024 * 1024 * 2
```
If we increase this number (by a factor of two for our 33 level 970*1042 UK data) we can step around the problem for the time being by stopping the Iris sub-chunking the slices to be merged. As here::
```
dask.array<array, shape=(33, 970, 1042), dtype=float32, chunksize=(16, 970, 1042)>
```
As we are using 32bit floats now, for UK data we would still have a chunk size of ~130MB , while global data would expand to ~650MB.

**Comment by LaurenceBeard** [comment]
This has now been fixed in https://github.com/dask/dask/issues/4043, we should be able to undo this fix this when the new Dask (>= 0.19.3) hits the stack.
